### ISSUE TITLE
[RFC] Fail-Closed Graph Storage Contract for Weight Reload

### ISSUE THREAD

**Issue by aoshen02** [vllm-project/vllm#48478]
## Summary

Introduce one explicit `GraphStorageRegistry` contract for tensors whose storage is retained by CUDA/HIP Graphs or backend objects across weight reloads.

The invariant is:

> If a weight update returns success, every registered graph-visible logical slot still resolves to the captured storage with a compatible layout. Otherwise the worker remains paused and the update fails.

This RFC is the production-design follow-up to #48312. Category 1 (storage identity) is enforced at runtime. Categories 2–4 continue to use the production post-load lifecycle plus the cold-B vs warm-A→B differential oracle defined in #48312.

## Confirmed source topology

Current `main` has two distinct update paths:

```text
Worker.reload_weights()
  -> ModelRunner.reload_weights()
  -> no WeightTransferEngine call

Worker.update_weights()
  -> WeightTransferEngine.update_weights()
```

Therefore the storage contract cannot live only in `WeightTransferEngine`. The ModelRunner owns the registry and the common completion check.

## Runtime assumptions

Weight reload runs after inference has been paused and drained. No inference executes during reload. Existing CUDA/HIP Graph objects remain alive and retain captured device addresses.

Consequences:

- reload updates the live model directly;
- no shadow model;
- no full A/B copy of model weights;
- no rollback to A;
- a failed reload leaves the worker paused/unhealthy and requires full reload or restart;
- no graph replay resumes before validation succeeds.

## Ownership

```text
Backend/layer
  -> register graph-visible storage

CUDA/HIP Graph capture boundary
  -> ModelRunner.storage_registry.snapshot(graph_id)

Direct reload or WeightTransferEngine finish
  -> ModelRunner.storage_registry.check_or_raise(graph_id)
  -> invalidate required caches
  -> commit weight generation
  -> allow scheduler resume
```

There is no direct Graph Manager ↔ WeightTransferEngine dependency.

## Single registration mechanism

Production uses one discovery mechanism: explicit registration.

```python
register_graph_storage(
    owner=layer,
    name="moe.experts.a_strides",
    tensor=a_strides,
    attr_path="quant_method.moe_kernel.fused_experts.a_strides",
)
```

- Parameters and buffers are inserted by a framework adapter through the same API.
- Backend-owned runtime tensors, workspaces, derived weights, and opaque resources are declared through a stable `nn.Module` owner plus an explicit attribute path. The resolver follows the live object chain after reload and never closes over the old kernel/tensor object.
- Graph capture snapshots the registry.
- Production does not use `TorchDispatchMode` discovery.
- Production does not recursively walk arbitrary Python object graphs.
- A backend without a complete reload-storage contract does not support warm reload while captured graphs remain alive; the operation fails closed.

Object walking remains an offline audit/CI tool. It generates review candidates and never defines the production manifest.

## Registry record

Each record contains:

- stable logical name;
- weak storage reference;
- storage identity / data pointer;
- shape, stride, storage offset;
- dtype and device;
- resolver for the current logical slot;
- graph generation identifier.

`id(tensor)` is insufficient because multiple tensor objects share storage and `.data = ...` replaces storage without replacing the Python object.

## Check semantics

After reload and before resume:

```python
current = record.resolver()

if current is None:
    raise ReloadStorageMismatch("logical slot disappeared")
if storage(current) is not record.captured_storage:
    raise ReloadStorageMismatch("logical slot rebound")
if layout(current) != record.captured_layout:
    raise ReloadStorageMismatch("layout changed")
```

The check does not compare values. Value correctness uses a value-changing A/B differential test:

```text
cold load B
==
load A -> capture -> reload B -> replay existing graph
```

## Completion gate

Both direct reload and weight-transfer updates enter the same ModelRunner-owned completion gate:

```text
accelerator synchronize
-> run existing PWAL / registered backend finalizers
-> storage_registry.check_or_raise()
-> invalidate required caches
-> commit weight_gen
-> resume
```

`process_weights_after_loading` (PWAL) remains the standard mechanism for recomputing packed, transposed, quantized, and other derived runtime values. A small explicit reload finalizer is reserved for non-standard backend/expert objects that are outside the existing PWAL lifecycle.

## Distributed failure

Every rank checks its local registry. Any rank failure fails the collective update. No rank commits the new generation and no scheduler resumes after a peer reports a mismatch.

## CI

1. Registry unit tests: alive storage, expired storage, slot rebind, layout mismatch, graph generations.
2. Entrypoint coverage: direct `reload_weights` and IPC/NCCL `finish_weight_update` both execute the completion gate.
3. Historical red/green tests: #48251, #41670, #46009, and #48438.
4. Hardware replay: capture A, reload B, allocation pressure, replay the existing graph, compare with cold B.
5. A source audit test reports post-load tensor creation/rebinding sites that lack registration. Audit output is a candidate list, not a runtime oracle.

## Rejected designs

- **WeightTransferEngine-owned registry:** rejected because direct `Worker.reload_weights()` bypasses WeightTransferEngine.
- **Automatic dispatch discovery:** rejected because compiled and opaque backend state is outside that contract.
- **Runtime arbitrary object walk:** rejected because reachability does not prove graph use and object graphs contain caches, handles, locks, cycles, and temporary state.
- **Check before every replay:** rejected; the reload completion boundary is sufficient and adds no steady-state overhead.
- **Automatic recapture:** rejected; it hides a violated reload contract.
- **Warning-only behavior:** rejected; stale or dangling graph addresses are correctness failures.
- **Shadow-model staging and rollback:** rejected because inference is quiesced during reload and failed workers do not resume.

## Implementation plan

1. Add `GraphStorageRegistry` and CPU unit tests.
2. Add ModelRunner ownership and capture snapshot lifecycle.
3. Add one common completion gate for direct reload and WeightTransferEngine completion.
4. Register parameters/buffers through the framework adapter.
5. Migrate known graph-visible backend state, starting with MLA, Marlin/Machete, CUTLASS MoE, FlashInfer B12x, and RDNA3 scratch.
6. Add GPU capture/reload/replay lanes.
7. Enable strict support only for backends whose contract is complete; unsupported combinations fail before mutating weights.

## Exit criteria

- A successful reload never resumes with a missing, rebound, or layout-incompatible registered graph storage.
- Unsupported captured-graph reload combinations fail before weight mutation.
- Direct reload and all WeightTransferEngine paths share the completion gate.
- Historical category-1 failures turn red without their fixes and green with them.
- Cold-B and warm-A→B existing-graph replay agree for the supported backend matrix.

AI assistance was used to audit the call paths and draft this RFC. I reviewed the design and source relationships and can defend the proposal end to end.
